ngraph/core/reference/src/runtime/reference/einsum.cpp

```diff
@@ -0,0 +1,1239 @@
+                                      size_t input_ind)
+                {
+                    // perform sanity check for arguments
+                    auto num_inputs = inputs.size();
+                    NGRAPH_CHECK(num_inputs == input_subscripts.size(),
+                                 "Each input must have own subscript.");
+                    NGRAPH_CHECK(input_ind < num_inputs, "Input index is out of range.");
+
+                    const auto& input_ptr = inputs[input_ind];
+                    const auto& input_subscript = input_subscripts[input_ind];
+                    const auto input_shape = input_ptr->get_shape();
+
+                    std::string resultant_subscript = "";
+                    constexpr char ellipsis[] = "...";
+                    auto labels = ngraph::opset7::Einsum::extract_labels(input_subscript);
+                    auto label_dim_map = compute_label_dim_map(
+                        input_ptr->get_partial_shape().rank(), input_subscript);
+                    std::vector<std::string> repeated_labels;
+                    Shape result_shape;
+                    AxisSet reduced_axes;
+                    for (const auto& label : labels)
+                    {
+                        if (resultant_subscript.find(label) == std::string::npos)
+                        {
+                            NGRAPH_CHECK(label_dim_map.find(label) != label_dim_map.end());
+                            auto dims = label_dim_map[label];
+                            NGRAPH_CHECK(dims.size() > 0);
+                            if (label != ellipsis && dims.size() > 1)
+                            {
+                                // repeated label is found
+                                for (size_t dim_ind = 1; dim_ind < dims.size(); ++dim_ind)
+                                {
+                                    reduced_axes.insert(dims[dim_ind]);
+                                }
+                                // save only the first dimension corresponding to the repeated label
+                                dims = {dims[0]};
+                                repeated_labels.push_back(label);
+                            }
+                            resultant_subscript += label;
+                            for (auto dim : dims)
+                            {
+                                NGRAPH_CHECK(dim < input_shape.size());
+                                result_shape.push_back(input_shape[dim]);
+                            }
+                        }
+                    }
+                    if (input_shape == result_shape)
+                    {
+                        return;
+                    }
+
+                    HostTensorPtr multi_identity =
+                        build_multi_identity<T>(input_ptr, repeated_labels, label_dim_map);
+
+                    HostTensorPtr mul_output = std::shared_ptr<HostTensor>(
```

Comment:
Looks like here we can use `input_ptr` as destination.

Reply:
Done.